aptos-move/framework/aptos-framework/sources/account/common_account_abstractions/ethereum_derivable_account.move

```diff
@@ -0,0 +1,429 @@
+        message.append(*domain);
+        message.append(b"\nVersion: 1");
+        message.append(b"\nChain ID: ");
+        message.append(*string_utils::to_string(&chain_id::get()).bytes());
+        message.append(b"\nNonce: ");
+        message.append(*digest_utf8);
+        message.append(b"\nIssued At: ");
+        message.append(*issued_at);
+
+        let msg_len = vector::length(message);
+
+        let prefix = b"\x19Ethereum Signed Message:\n";
+        let msg_len_string = string_utils::to_string(&msg_len); // returns string
+        let msg_len_bytes = msg_len_string.bytes(); // vector<u8>
+
+        let full_message = &mut vector[];
+        full_message.append(prefix);
+        full_message.append(*msg_len_bytes);
+        full_message.append(*message);
+
+        aptos_hash::keccak256(*full_message)
+    }
+
+    fun recover_public_key(signature_bytes: &vector<u8>, message: &vector<u8>): vector<u8> {
+        let rs = vector::slice(signature_bytes, 0, 64);
+        let v = *vector::borrow(signature_bytes, 64);
+        assert!(v == 27 || v == 28, EUNEXPECTED_V);
+        let signature = secp256k1::ecdsa_signature_from_bytes(rs);
+
+        let maybe_recovered = secp256k1::ecdsa_recover(*message, v - 27, &signature);
+
+        assert!(
+            option::is_some(&maybe_recovered),
+            EINVALID_SIGNATURE
+        );
+
+        let pubkey = option::borrow(&maybe_recovered);
+
+        let pubkey_bytes = secp256k1::ecdsa_raw_public_key_to_bytes(pubkey);
+
+        // Add 0x04 prefix to the public key, to match the
+        // full uncompressed format from ethers.js
+        let full_pubkey = &mut vector[];
+        vector::push_back(full_pubkey, 4u8);
+        vector::append(full_pubkey, pubkey_bytes);
+
+        *full_pubkey
+    }
+
+    fun entry_function_name(entry_function_payload: &EntryFunctionPayload): vector<u8> {
+        let entry_function_name = &mut vector[];
+        let addr_str = string_utils::to_string(
+            &transaction_context::account_address(entry_function_payload)
+        ).bytes();
+        // .slice(1) to remove the leading '@' char
+        entry_function_name.append(addr_str.slice(1, addr_str.length()));
+        entry_function_name.append(b"::");
+        entry_function_name.append(
+            *transaction_context::module_name(entry_function_payload).bytes()
+        );
+        entry_function_name.append(b"::");
+        entry_function_name.append(
+            *transaction_context::function_name(entry_function_payload).bytes()
+        );
+        *entry_function_name
+    }
+
+
+
+    fun hex_char_to_u8(c: u8): u8 {
+        if (c >= 48 && c <= 57) {  // '0' to '9'
+            c - 48
+        } else if (c >= 65 && c <= 70) { // 'A' to 'F'
+            c - 55
+        } else if (c >= 97 && c <= 102) { // 'a' to 'f'
+            c - 87
+        } else {
+            abort 1
+        }
+    }
+
+    public fun base16_utf8_to_vec_u8(str: vector<u8>): vector<u8> {
```

Comment:
@igor-aptos do you know if we have a framework function we can use here instead?

Reply:
probably not, worth asking on #move-eng , of whether it should be put in aptos_stdlib

any reason for this to be public?

Reply:
>  any reason for this to be public?

No